pkg/ring/util.go

```diff
@@ -141,13 +141,25 @@ func waitStability(ctx context.Context, r *Ring, op Operation, minStability, max
 }
 
 // MakeBuffersForGet returns buffers to use with Ring.Get().
-func MakeBuffersForGet() (bufDescs []InstanceDesc, bufHosts, bufZones []string) {
+func MakeBuffersForGet() (bufDescs []InstanceDesc, bufHosts []string, bufZones map[string]int) {
 	bufDescs = make([]InstanceDesc, 0, GetBufferSize)
 	bufHosts = make([]string, 0, GetBufferSize)
-	bufZones = make([]string, 0, GetBufferSize)
+	bufZones = make(map[string]int, GetZoneSize)
 	return
 }
 
+func resetZoneMap(zones map[string]int) map[string]int {
+	if zones == nil {
+		return make(map[string]int, GetZoneSize)
+	}
+
+	for key := range zones {
```

Comment:
Assuming we have this optimization to avoid allocating too many objects in the heap, is it better to reset the values to 0?

Reply:
Some of the components call `ring.Get` func with nil passed to `bufDesc`, `bufHosts` and `bufZones`. In those cases, it's expected for the arrays and map to be created within the Get function. Hence I'm checking for nil map in `resetZoneMap` and creating a new one if true.

In cases where `ring.Get` func is called a lot of times, yes we use the `buf*` variables to avoid allocation. We also have a test `TestRing_Get_NoMemoryAllocations` to verify no allocation is happening in `ring.Get`.

As we initially specify the size of the buffer map as `make(map[string]int, GetZoneSize)`, I don't think there will be any difference in deleting the map entry or just setting the value of 0. However, the benefits of deleting the map entry are:

1. We don't have to worry whether the ring zone info has changed
2. `len(zones)` will reflect how many zones have at least one instance